compiler/packages/babel-plugin-react-compiler/src/HIR/visitors.ts

```diff
@@ -1149,3 +1153,72 @@ export function* eachTerminalOperand(terminal: Terminal): Iterable<Place> {
     }
   }
 }
+
+/**
+ * Helper class for traversing scope blocks in HIR-form.
+ */
+export class ScopeBlockTraversal {
+  // Live stack of active scopes
+  #activeScopes: Array<ScopeId> = [];
+  blockInfos: Map<
+    BlockId,
+    | {
+        kind: 'end';
+        scope: ReactiveScope;
+        pruned: boolean;
+      }
+    | {
+        kind: 'begin';
+        scope: ReactiveScope;
+        pruned: boolean;
+        fallthrough: BlockId;
+      }
+  > = new Map();
+
+  recordScopes(block: BasicBlock): void {
+    const blockInfo = this.blockInfos.get(block.id);
+    if (blockInfo?.kind === 'begin') {
+      this.#activeScopes.push(blockInfo.scope.id);
+    } else if (blockInfo?.kind === 'end') {
+      const top = this.#activeScopes.at(-1);
+      CompilerError.invariant(blockInfo.scope.id === top, {
+        reason:
+          'Expected traversed block fallthrough to match top-most active scope',
+        loc: block.instructions[0]?.loc ?? block.terminal.id,
+      });
+      this.#activeScopes.pop();
+    }
+
+    if (
+      block.terminal.kind === 'scope' ||
+      block.terminal.kind === 'pruned-scope'
+    ) {
+      CompilerError.invariant(
+        !this.blockInfos.has(block.terminal.block) &&
+          !this.blockInfos.has(block.terminal.fallthrough),
+        {
+          reason: 'Expected unique scope blocks and fallthroughs',
+          loc: block.terminal.loc,
+        },
+      );
+      this.blockInfos.set(block.terminal.block, {
+        kind: 'begin',
+        scope: block.terminal.scope,
+        pruned: block.terminal.kind === 'pruned-scope',
+        fallthrough: block.terminal.fallthrough,
+      });
+      this.blockInfos.set(block.terminal.fallthrough, {
+        kind: 'end',
+        scope: block.terminal.scope,
+        pruned: block.terminal.kind === 'pruned-scope',
+      });
+    }
+  }
+
+  isScopeActive(scopeId: ScopeId): boolean {
+    return this.#activeScopes.indexOf(scopeId) !== -1;
```

Comment:
I believe the `indexOf` check should be cheap, as # of nested scopes is generally low. Happy to add a backing set if this isn't true though